File: project_files/Android-build/SDL-android-project/src/org/hedgewars/hedgeroid/Datastructures/Scheme.java
/*
 * Hedgewars for Android. An Android port of Hedgewars, a free turn based strategy game
 * Copyright (C) 2012 Simeon Maxein <smaxein@googlemail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.hedgewars.hedgeroid.Datastructures;

import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

public final class Scheme {
	public final String name;
	public final Map<String, Integer> settings;
	public final Map<String, Boolean> mods;
		
	public Scheme(String name, Map<String, Integer> settings, Map<String, Boolean> mods) {
		this.name = name;
		this.settings = Collections.unmodifiableMap(new HashMap<String, Integer>(settings));
		this.mods = Collections.unmodifiableMap(new HashMap<String, Boolean>(mods));
	}
	
	public int getHealth() {
		Integer health = settings.get("health");
		return health==null ? 100 : health.intValue();
	}

	public static Scheme createDefaultScheme(MetaScheme meta) {
		String name = GameConfig.DEFAULT_SCHEME;
		Map<String, Integer> settings = new TreeMap<String, Integer>();
		Map<String, Boolean> mods = new TreeMap<String, Boolean>();
		for(MetaScheme.Setting setting : meta.settings) {
			settings.put(setting.name, setting.def);
		}
		for(MetaScheme.Mod mod : meta.mods) {
			mods.put(mod.name, Boolean.FALSE);
		}
		return new Scheme(name, settings, mods);
	}
	
	@Override
	public String toString() {
		return "Scheme [name=" + name + ", settings=" + settings + ", mods="
				+ mods + "]";
	}
	
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((name == null) ? 0 : name.hashCode());
		result = prime * result + ((mods == null) ? 0 : mods.hashCode());
		result = prime * result
				+ ((settings == null) ? 0 : settings.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Scheme other = (Scheme) obj;
		if (name == null) {
			if (other.name != null)
				return false;
		} else if (!name.equals(other.name))
			return false;
		if (mods == null) {
			if (other.mods != null)
				return false;
		} else if (!mods.equals(other.mods))
			return false;
		if (settings == null) {
			if (other.settings != null)
				return false;
		} else if (!settings.equals(other.settings))
			return false;
		return true;
	}

	public static final Comparator<Scheme> NAME_ORDER = new Comparator<Scheme>() {
		public int compare(Scheme lhs, Scheme rhs) {
			return String.CASE_INSENSITIVE_ORDER.compare(lhs.name, rhs.name);
		}
	};
}

File: project_files/frontlib/model/scheme.c
/*
 * Hedgewars, a free turn based strategy game
 * Copyright (C) 2012 Simeon Maxein <smaxein@googlemail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include "scheme.h"

#include "../util/inihelper.h"
#include "../util/logging.h"
#include "../util/util.h"

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>

flib_scheme *flib_scheme_create(const char *schemeName) {
    flib_scheme *result = flib_calloc(1, sizeof(flib_scheme));
    if(log_badargs_if(schemeName==NULL) || result==NULL) {
        return NULL;
    }

    result->name = flib_strdupnull(schemeName);
    result->mods = flib_calloc(flib_meta.modCount, sizeof(*result->mods));
    result->settings = flib_calloc(flib_meta.settingCount, sizeof(*result->settings));

    if(!result->mods || !result->settings || !result->name) {
        flib_scheme_destroy(result);
        return NULL;
    }

    for(int i=0; i<flib_meta.settingCount; i++) {
        result->settings[i] = flib_meta.settings[i].def;
    }
    return result;
}

flib_scheme *flib_scheme_copy(const flib_scheme *scheme) {
    flib_scheme *result = NULL;
    if(scheme) {
        result = flib_scheme_create(scheme->name);
        if(result) {
            memcpy(result->mods, scheme->mods, flib_meta.modCount * sizeof(*scheme->mods));
            memcpy(result->settings, scheme->settings, flib_meta.settingCount * sizeof(*scheme->settings));
        }
    }
    return result;
}

void flib_scheme_destroy(flib_scheme* scheme) {
    if(scheme) {
        free(scheme->mods);
        free(scheme->settings);
        free(scheme->name);
        free(scheme);
    }
}

bool flib_scheme_get_mod(const flib_scheme *scheme, const char *name) {
    if(!log_badargs_if2(scheme==NULL, name==NULL)) {
        for(int i=0; i<flib_meta.modCount; i++) {
            if(!strcmp(flib_meta.mods[i].name, name)) {
                return scheme->mods[i];
            }
        }
        flib_log_e("Unable to find game mod %s", name);
    }
    return false;
}

int flib_scheme_get_setting(const flib_scheme *scheme, const char *name, int def) {
    if(!log_badargs_if2(scheme==NULL, name==NULL)) {
        for(int i=0; i<flib_meta.settingCount; i++) {
            if(!strcmp(flib_meta.settings[i].name, name)) {
                return scheme->settings[i];
            }
        }
        flib_log_e("Unable to find game setting %s", name);
    }
    return def;
}
